### PR TITLE
interface: convert stores to zustand

### PR DESCRIPTION
**Pull request by tylershuster**
This PR moves all the stores to zustand. I will try to make a separate commit for each store, which should implement the store, convert the reducer, and strip all the drilled props belonging to that store. Each commit will build on the last and should compile completely.

- [x] contact
- [x] graph
- [x] groups
- [x] hark
- [x] invite
- [x] launch
- [x] metadata
- [x] ~s3~ storage
- [x] settings

### CODE REVIEW

File: pkg/interface/src/logic/state/contacts.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  //   ship,
+  //   personal
+  // ): Promise<boolean> => {
+  //   const isPersonal = personal ? 'true' : 'false';
+  //   const api = useApi();
+  //   return api.scry({
+  //     app: 'contact-store',
+  //     path: `/is-allowed/${entity}/${name}/${ship}/${isPersonal}`
+  //   });
+  // },
```
Comment (tylershuster) [inline pkg/interface/src/logic/state/contacts.tsx:34]: trust the plan

**Comment by tylershuster** [comment]
feelsgoodman.jpg

**Comment by tylershuster** [comment]
**Note** (this should be its own issue) — non-serializable types such as `Set<string>` such as `graphKeys` cannot be persisted in localStorage and will therefore be ignored **for persistence** in this pass. We should write a custom storage handler for each store to properly dehydrate these from localStorage. The current solution will not offer any performance benefits for those stores, but it will also not inflict any performance hits

**Comment by tylershuster** [comment]
Alright, this has some issues with immutability that we need to address but I think generally it's ready to review

**Comment by tylershuster** [comment]
Ok, there's definitely something weird going on after the initial population of the store, likely related to the use of immer in the setter. I'm working on a reusable (and also usable) pattern for reducing state.

**Comment by tylershuster** [comment]
Now the stores actually work, but at the cost of losing immutability. I don't think we're ever actually using the benefits of `immer`, so I'm not sad but I would like to figure out how to do this properly. The relevant method is in `lib/util/reduceState`

**Comment by tylershuster** [comment]
@liam-fitzgerald We should talk at some point about the methods I'm using to weed out any potential problems before I convert `settings` to use the same pattern

**Review by liam-fitzgerald — APPROVED** [review]

**Comment by tylershuster** [comment]
Copying from out of band:

- [ ] Converting settings store
- [ ] I forgot to do GroupViewReducer
- [ ] localreducer?
- [ ] the connection handling I don't quite understand
- [ ] implementing the debugStore
- [ ] deleting all the old references to the global store members

**Comment by matildepark** [comment]
All these todos and somehow this got an approve stamp.

**Comment by tylershuster** [comment]
He talked to me after the stamp. It won't break anything if it goes in, just needs some cleaning up

**Comment by tylershuster** [comment]
So sassy

**Comment by matildepark** [comment]
Hmm. Merge next-js back in?

**Comment by tylershuster** [comment]
Ok, but don't merge this until you've built it. There's a weird error that *shouldn't* be happening, but is.

**Comment by tylershuster** [comment]
To be exceedingly clear, there are a few niggling things in here which should be cleaned up (like the base hash and connection state) but are better addressed once the API is switched out.

File: pkg/interface/src/logic/lib/withState.tsx [hunk 2]
```diff
@@ -20,4 +20,27 @@ const withState = <
...
+          keys.reduce(
+            (object, key) => ({ ...object, [key]: state[key] }), {}
+          )
+        })
+        : store();
+      Object.assign(stateProps, storeProps);
+    })
+    return <Component ref={ref} {...stateProps} {...props} />
+  });
+}
```
Comment (tylershuster) [inline pkg/interface/src/logic/lib/withState.tsx:44]: @liam-fitzgerald Clear as mud?

**Comment by tylershuster** [comment]
I will resolve these conflicts once stamped

File: pkg/interface/src/logic/lib/withState.tsx (hunk 2 above)
Comment (liam-fitzgerald) [reply pkg/interface/src/logic/lib/withState.tsx:44]: Nah that's not that bad

File: pkg/interface/src/views/apps/graph/app.js [hunk 3]
```diff
@@ -2,58 +2,53 @@ import React, { PureComponent } from 'react';
...
-            } else if(!!association) {
-              props.history.push(`/~landscape/home/resource/${association.metadata.module}${path}`);
+import useGraphState from '~/logic/state/graph';
+import useMetadataState from '~/logic/state/metadata';
+
+const GraphApp = (props) => {
+  const associations= useMetadataState(state => state.associations);
+  const graphKeys = useGraphState(state => state.graphKeys);
+
+  const { api } = this.props;
```
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/graph/app.js:12]: hm??

File: pkg/interface/src/views/landscape/components/JoinGroup.tsx [hunk 4]
```diff
@@ -75,12 +77,12 @@ export function JoinGroup(props: JoinGroupProps): ReactElement {
     await api.groups.join(ship, name);
     try {
       await waiter((p: JoinGroupProps) => {
```
Comment (liam-fitzgerald) [inline pkg/interface/src/views/landscape/components/JoinGroup.tsx:79]: You're going to want to pass the hook values into `useWaitForProps` and then check there

**Review by liam-fitzgerald — CHANGES_REQUESTED** [review]
Did a full read of this. Looks good. Obviously needs an update for the new GCP storage stuff, merge conflicts and comments need to be resolved.

File: pkg/interface/src/views/apps/graph/app.js (hunk 3 above)
Comment (tylershuster) [reply pkg/interface/src/views/apps/graph/app.js:12]: hm hm? All of these are used in the file.

**Comment by tylershuster** [comment]
@liam-fitzgerald  Can you give this a run? I'm getting this weird error that seems to be a result of something failing to import correctly, but I see no reason why it wouldn't.

**Comment by liam-fitzgerald** [comment]
don't see any issues when building this, other than a missing quote on L110 of `lib/tutorialModal.ts`

**Comment by tylershuster** [comment]
Alright, I will wipe and reload. Once I've built it and checked I'll make a final commit wrapping this up (God willing).

**Comment by tylershuster** [comment]
@liam-fitzgerald discovered a circular dependency that was causing the build to fail. Once I fixed that I discovered a few other problems clicking around the interface. Now it works and in my limited testing I haven't been able to break it. I think this is ready for testnet

**Comment by tylershuster** [comment]
one day

**Comment by matildepark** [comment]
Conflicts again. Otherwise we can start shoving it into next-js once it has an all clear and we reasonably expect it to work.

**Comment by tylershuster** [comment]
I'm under the impression it has an all-clear, I just don't want to keep cleaning up conflicts. @liam-fitzgerald ?

**Comment by liam-fitzgerald** [comment]
merge, and then i'll test. If works, i'll merge

**Comment by matildepark** [comment]
We were basically holding bc next-js was going into the next release (today) and this needs some dev stream time first. But this blocks a lot of big ticket so it should be expedited if we have baseline confidence in its stability.

**Comment by tylershuster** [comment]
aye aye. I would not be comfortable with this going out today; needs a little bit of field experience first.

**Comment by tacryt-socryp** [comment]
I'd love to get this in release/next-js so we can test ASAP

**Comment by tacryt-socryp** [comment]
Are we planning another release today, or are you referring to the earlier one @matildepark ?

**Comment by liam-fitzgerald** [comment]
posting a link comment crashes

**Comment by matildepark** [comment]
Earlier. If we release today it will be hotfix based on the remaining p1s on Liam’s plate. 

— 
~haddef-sigwen
https://urbit.org

**Review by liam-fitzgerald — APPROVED** [review]
LGTM!

**Pull request merged**